src/flake8_numbers/check_numbers.py

```diff
@@ -0,0 +1,209 @@
+
+
+@dataclass
+class ErrorReport:
+    """A class to represent an error report."""
+
+    line: int
+    """The line number of the error."""
+
+    column: int
+    """The column number of the error."""
+
+    message: str
+    """The error message."""
+
+
+def _base_value(number_literal: str) -> int:
+    """Get the base value of the given number literal.
+
+    Args:
+        number_literal: The number literal to get the base value of.
+
+    Returns:
+        The base value of the given number literal.
+    """
+    lower_number_literal = number_literal.lower()
+    if lower_number_literal.startswith("0b"):
+        return 2
+    if lower_number_literal.startswith("0o"):
+        return 8
+    if lower_number_literal.startswith("0x"):
+        return 16
+    return 10
+
+
+def _separator_modulo_for_base(base: int) -> int:
+    """Get the separator modulo for the given base.
+
+    Note: We keep this currently as helper function, to make that part of the code
+    more readable as well as open up the possibility of making this configurable
+    in the future.
+
+    Args:
+        base: The base to get the separator modulo for.
+
+    Returns:
+        The separator modulo for the given base.
+    """
+    match base:
+        case 2:
+            return 4
+        case 8:
+            return 4
+        case 16:
+            return 4
+        case _:
+            return 3
+
+
+class Flake8NumbersChecker:
+    """class to represent a flake8 plugin to check for numbers and their readability."""
+
+    name = "flake8.numbers"
+    version = metadata.version("flake8-numbers")
+    off_by_default = False
+
+    # Important: The parameter names must match exactly the way how flake8 expects them.
+    # This is sadly undocumented and we only found out by looking into the source code.
+    # But it is what it is.
+    def __init__(self, tree: ast.AST, filename: str) -> None:
+        """Initialize the checker.
+
+        Args:
+            tree: The AST of the file being checked.
+            filename: The path to the file being checked.
+        """
+        self._tree = tree
+        self._filename = filename
+        self._source_lines: Optional[list[str]] = None
+
+    @property
+    def source_lines(self) -> list[str]:
+        """Get the source lines of the file being checked.
+
+        This value is cached on-demand.
+
+        Returns:
+            The source lines of the file being checked.
+        """
+        if self._source_lines is None:
+            with open(self._filename, "r", encoding="utf-8") as source_file:
+                self._source_lines = source_file.readlines()
+        return self._source_lines
+
+    def run(self) -> Iterable[Tuple[int, int, str, Type[Any]]]:
+        """Run the checker.
+
+        Yields:
+            A tuple of the form (line, column, message, type).
+        """
+        for node in ast.walk(self._tree):
+            if isinstance(node, ast.Num):
```

Comment:
Mmm, this doesn't look good:

> Deprecated since version 3.8: Old classes ast.Num, ast.Str, ast.Bytes, ast.NameConstant and ast.Ellipsis are still available, but they will be removed in future Python releases. In the meantime, instantiating them will return an instance of a different class.

https://docs.python.org/3/library/ast.html#ast.AST